src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIAssistantClient.cs

```diff
@@ -0,0 +1,379 @@
+        }
+        else
+        {
+            updates = _assistantClient.CreateRunStreamingAsync(threadId, assistantId, runOptions, cancellationToken);
+        }
+
+        // Process each update.
+        await foreach (var update in updates.ConfigureAwait(false))
+        {
+            switch (update)
+            {
+                case MessageContentUpdate mcu:
+                    yield return new()
+                    {
+                        RawRepresentation = mcu,
+                        Role = mcu.Role == MessageRole.User ? ChatRole.User : ChatRole.Assistant,
+                        Text = mcu.Text,
+                    };
+                    break;
+
+                case ThreadUpdate tu when options is not null:
+                    options.ConversationId ??= tu.Value.Id;
+                    break;
+
+                case RunUpdate ru:
+                    StreamingChatCompletionUpdate ruUpdate = new()
+                    {
+                        AuthorName = ru.Value.AssistantId,
+                        CompletionId = ru.Value.Id,
+                        CreatedAt = ru.Value.CreatedAt,
+                        ModelId = ru.Value.Model,
+                        RawRepresentation = ru,
+                        Role = ChatRole.Assistant,
+                    };
+
+                    if (ru.Value.Usage is { } usage)
+                    {
+                        ruUpdate.Contents.Add(new UsageContent(new()
+                        {
+                            InputTokenCount = usage.InputTokenCount,
+                            OutputTokenCount = usage.OutputTokenCount,
+                            TotalTokenCount = usage.TotalTokenCount,
+                        }));
+                    }
+
+                    if (ru is RequiredActionUpdate rau && rau.ToolCallId is string toolCallId && rau.FunctionName is string functionName)
+                    {
+                        ruUpdate.Contents.Add(
+                            new FunctionCallContent(
+                                toolCallId, functionName,
+                                JsonSerializer.Deserialize(rau.FunctionArguments, OpenAIJsonContext.Default.IDictionaryStringObject)!)
+                            {
+                                Context = rau.GetThreadRun().Id
+                            });
+                    }
+
+                    break;
+            }
+        }
+    }
+
+    /// <inheritdoc />
+    void IDisposable.Dispose()
+    {
+        // Nothing to dispose. Implementation required for the IChatClient interface.
+    }
+
+    /// <summary>Gets the thread ID and assistant ID to target.</summary>
+    private (string? AssistantId, string? ThreadId) GetIds(ChatOptions? options)
+    {
+        // Get the assistant ID to use.
+        string? assistantId = options?.AdditionalProperties?.TryGetValue(AssistantIdKey, out string? s) is true ? s : _assistantId;
```

Comment:
I would personally prefer it if there was only one way of specifying these id's, ideally from the constructor. The `ChatOptions` could have been deserialized from the wire, so in a hypothetical agent implementation these could be overridden by callers.

Reply:
We already allow things like model and dimensions to be specified in multiple places, with one baked into the instance and options used for ovverides or if one wasn't specified to construction. And in practice, there's nothing we can do to avoid that... anyone can define an IChatClient with any defaults built into it for things otherwise specified by ChatOptions.